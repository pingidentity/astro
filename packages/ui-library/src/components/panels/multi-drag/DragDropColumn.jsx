var React = require("re-react"),
    ReactDOM = require("react-dom"),
    FormSearchBox = require("../../forms/FormSearchBox.jsx"),
    FormLabel = require("../../forms/FormLabel.jsx"),
    DDRow = require("../../rows/DragDropRow.jsx"),
    classnames = require("classnames"),
    _ = require("underscore");

/**
 * @class MultiDrag#DragDropColumn
 * @private
 * @ignore
 *
 * @desc This component is internal to the MultiDrag component and is used to render a column of drag/drop rows.
 *
 * @param {string} [data-id="drag-drop-column"]
 *    To define the base "data-id" value for the top-level HTML container.
 *
 * @param {object[]} rows
 *    The row objects. The shape of these objects must correspond to the expected props of your contentType.
 * @param {string} name
 *    The name of the column.
 * @param {number} index
 *    The index of the column. Used when executing callbacks.
 * @param {string} filter
 *    The filter to display in the search field.
 * @param {object} contentType
 *    A react component to be used as a template for rendering rows.
 * @param {string} [labelEmpty]
 *    The placeholder string for when a column is empty.
 * @param {boolean} [showSearch=false]
 *    Determines if the searchbox should be shown.
 * @param {string} [className]
 *    CSS classes to set on the top-level HTML container.
 * @param {number} [ghostRowAt]
 *    Display a ghost row to preview where a row would be inserted
 * @param {boolean} [disableSort=false]
 *    Determines whether the column in internally sortable
 *
 * @param {MultiDrag~onSearch} onSearch
 *    Callback to be triggered when a column is searched.
 * @param {MultiDrag~onDragDrop} onDrag
 *    Callback to be triggered when a row is dragged.
 * @param {MultiDrag~onDragDrop} onDrop
 *    Callback to be triggered when a row id dropped.
 * @param {MultiDrag~onCancel} onCancel
 *    Callback to be triggered when a drag event ends.
 * @param {MultiDrag~onScrolledToPosition} [onScrolledToTop]
 *    Callback to be triggered when the list is scrolled to the top.
 * @param {MultiDrag~onScrolledToPosition} [onScrolledToBottom]
 *    Callback to be triggered when the list is scrolled to the bottom. Can be used to fetch more data.
 *
 * @param {MultiDrag~onDragStart} [onDragStart]
 *    Callback to be triggered when a drag event starts.
 * @param {MultiDrag~onDragEnd} [onDragEnd]
 *    Callback to be triggered when a drag even ends.
 *
 */
module.exports = React.createClass({
    propTypes: {
        "data-id": React.PropTypes.string,
        rows: React.PropTypes.arrayOf(
            React.PropTypes.object
        ).isRequired.affectsRendering,
        name: React.PropTypes.string.isRequired.affectsRendering,
        index: React.PropTypes.number.isRequired.affectsRendering,
        filter: React.PropTypes.string.affectsRendering,
        contentType: React.PropTypes.element.isRequired.affectsRendering,
        labelEmpty: React.PropTypes.string.affectsRendering,
        // optional
        showSearch: React.PropTypes.bool.affectsRendering,
        ghostRowAt: React.PropTypes.number.affectsRendering,
        className: React.PropTypes.string.affectsRendering,
        disableSort: React.PropTypes.bool.affectsRendering,
        // callbacks
        onSearch: React.PropTypes.func.isRequired,
        onDrag: React.PropTypes.func.isRequired,
        onDrop: React.PropTypes.func.isRequired,
        onCancel: React.PropTypes.func.isRequired,
        onScrolledToBottom: React.PropTypes.func,
        onScrolledToTop: React.PropTypes.func,
        onDragStart: React.PropTypes.func,
        onDragEnd: React.PropTypes.func
    },

    getDefaultProps: function () {
        return {
            "data-id": "drag-drop-column",
            showSearch: false,
            onScrolledToBottom: _.noop,
            onScrolledToTop: _.noop,
            labelEmpty: "No Items Added",
            onDragStart: _.noop,
            onDragEnd: _.noop
        };
    },

    /*
     * Since the multi-Drag component supports search in all columns, each column needs to report which
     * column a search affects.
     */
    _handleSearch: function (value) {
        this.props.onSearch(this.props.index, value);
    },

    /*
     * Handler for scrolls.  This is useful for implementing lazy loading of rows
     */
    _handleScroll: function () {
        var container = ReactDOM.findDOMNode(this.refs.items);
        var rect = container.getBoundingClientRect();

        if (container.scrollTop === 0) {
            this.props.onScrolledToTop(this.props.index);
        } else if (container.scrollTop + rect.height === container.scrollHeight) {
            this.props.onScrolledToBottom(this.props.index);
        }
    },

    /*
     * Renders the searchbar if the prop is set
     */
    _renderSearch: function () {
        if (!this.props.showSearch) {
            return null;
        }

        return (
            <div className="input-row" data-id="search">
                <FormSearchBox
                    onValueChange={this._handleSearch}
                    queryString={this.props.filter}
                    className="input-search"/>
            </div>);
    },

    /*
     * Renders a drag/drop from give a data object
     */
    _renderRow: function (row, index, opts) {
        var inner = opts && opts.content;

        if (!inner) {
            var props = _.extend({ column: this.props.index, index: index }, row);
            inner = React.cloneElement(this.props.contentType, props);
        }

        return (
            <DDRow id={row.id} key={row.id} index={index}
                   disabled={opts && opts.disabled}
                   column={this.props.index}
                   onDrag={this.props.onDrag}
                   onDrop={this.props.onDrop}
                   onDragStart={this.props.onDragStart}
                   onDragEnd={this.props.onDragEnd}
                   onCancel={this.props.onCancel}>
            {
                inner
            }
            </DDRow>);
    },

    _renderRows: function () {
        if (!this.props.rows || this.props.rows.length === 0) {
            var className = classnames("no-items", {
                preview: typeof(this.props.ghostRowAt) === "number"
            });

            return this._renderRow({ id: 0 }, 0, {
                content: <div data-id="empty-placeholder" className={className}>{this.props.labelEmpty}</div>,
                disabled: true
            });
        }

        var rows = this.props.rows.map(this._renderRow);

        // if there's a preview add it after so it doesnt screw up the indexing
        if (typeof(this.props.ghostRowAt) === "number") {
            var index = Math.min(this.props.rows.length, this.props.ghostRowAt);

            rows.splice(index, 0, this._renderRow({
                preview: true,
                id: "preview"
            }, index));
        }

        return rows;
    },

    render: function () {
        
        var className = classnames(
            this.props.className, {
                "disable-sort": this.props.disableSort
            }
        );
        
        return (
            <div data-id={this.props["data-id"]} className={className}>
                <FormLabel value={this.props.name} />

                { this._renderSearch() }

                <div className="items" onScroll={this._handleScroll} ref="items">
                    { this._renderRows() }
                </div>
            </div>);
    }
});
