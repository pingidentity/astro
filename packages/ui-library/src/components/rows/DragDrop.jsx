"use strict";

var React = require("react");
var ReactDOM = require("react-dom");
var dnd = require("react-dnd");
var classnames = require("classnames");
var dragSource = dnd.DragSource;
var dropTarget = dnd.DropTarget;
var PropTypes = React.PropTypes;
var TYPE = "DragDrop";
var _= require("underscore");

var dragSpec = {
    beginDrag: function (props) {
        return {
            id: props.id,
            index: props.index,
            column: props.column
        };
    },

    endDrag: function (props, monitor, component) {//eslint-disable-line
        if (props.id === monitor.getItem().id) {
            props.onCancel();
        }
    },

    canDrag: function (props) {
        return !props.disabled;
    }
};

var dragCollect = function (connect, monitor) {
    return {
        connectDragSource: connect.dragSource(),
        isDragging: monitor.isDragging()
    };
};

var isInTopHalf = function (monitor, node) {
    var boundingRect = node.getBoundingClientRect();
    var clientOffset = monitor.getClientOffset();
    var ownMiddleY = (boundingRect.bottom - boundingRect.top) / 2;

    return (clientOffset.y - boundingRect.top) <= ownMiddleY;
};
var isInLeftHalf = function (monitor, node) {
    var boundingRect = node.getBoundingClientRect();
    var clientOffset = monitor.getClientOffset();
    var ownMiddleX = (boundingRect.right - boundingRect.left) / 2;

    return (clientOffset.x - boundingRect.left) <= ownMiddleX;
};

function handleDragEvent (callback, props, monitor, component) {
    var itemBeingDragged = monitor.getItem();
    var locationType = props.type === "column" ? isInLeftHalf : isInTopHalf;
    var offset = locationType(monitor, ReactDOM.findDOMNode(component)) ? 0 : 1;
    //this is awkward but backward compatible
    props[callback](props.index + offset, itemBeingDragged.index, props.column, itemBeingDragged.column);
}

var dropSpec = {
    /*
     * @param {object} props - the properties of the target row
     * @param {object} monitor - the monitor object:
     *  - monitor.getItem() will return the row being dragged
     *  - monitor.getClientOffset() will return the mouse position.
     * @param {object} component - the target row react instance
     * @returns null - nothing
     */
    hover: handleDragEvent.bind(null, "onDrag"),
    drop: handleDragEvent.bind(null, "onDrop")
};

var dropCollect = function (connect) {
    return {
        connectDropTarget: connect.dropTarget()
    };
};

/**
 * @callback DragDrop~onDrag
 * @param {number} targetIndex
 *          current target index where items is dragged over or dropped
 * @param {number} sourceIndex
 *          origin index item is dragged or dropped from
 */

/**
 * @callback DragDrop~onDrop
 * @param {number} targetIndex
 *          current target index where items is dragged over or dropped
 * @param {number} sourceIndex
 *          origin index item is dragged or dropped from
 */

/**
 * @callback DragDrop~onCancel
 */

/**
 * @class DragDrop
 *
 * @desc A component to hide the workings of react-dnd.  Can be dragged on dropped onto any other component
 * of type DragDrop
 *
 * @param {string|number} id
 *          An id for the row
 * @param {number} index
 *          The index number for the row
 * @param {bool} [disabled=false]
 *          If child content should not be draggable. False by default (draggable).
 * @param {boolean} removeDraggableAttribute
 *          Remove the draggable="true" attribute to disable dragging.  This is useful to resolve issues with IE
 *          and input fields.
 * @param {DragDrop~onDrag} onDrag
 *          A callback which is execute when the row is moved.  Must have signature function (targetIndex, ownIndex)
 * @param {DragDrop~onDrop} onDrop
 *          A callback which is execute when the row is dropped.  Must have signature function (targetIndex, ownIndex)
 * @param {DragDrop~onCancel} onCancel
 *          A callback which is executed when the dragging is cancelled (dropped outside a droppable area
 *          or esc button pressed).
 * @param {string} [type]
 *          String to specify whether the component is a row or a column. Default "row".
 * @param {string} [tagName]
 *          The DOM node to wrap the draggable component. Default "div"
 * @param {object} [style]
 *          Inline style object for child
 * @example
 *
 *  <Draggable id={index} index={index} onDrag={this._onDrag} onDrop={this._onDrop}
 *              onCancel={this._onCancel} disabled={item === 2}>
 *      <div className={this._getDragClass(item, index)}>Row {item} {item === 2 && "can't drag me"}</div>
 *  </Draggable>
 *
 */
var DragDrop = React.createClass({
    propTypes: {
        connectDragSource: PropTypes.func.isRequired,
        connectDropTarget: PropTypes.func.isRequired,
        isDragging: PropTypes.bool.isRequired,

        id: PropTypes.any.isRequired,
        index: PropTypes.number.isRequired,
        onDrag: PropTypes.func.isRequired,
        onDrop: PropTypes.func.isRequired,
        onCancel: PropTypes.func.isRequired,

        type: PropTypes.string,
        tagName: PropTypes.string,
        style: PropTypes.object,

        disabled: PropTypes.bool,
        removeDraggableAttribute: PropTypes.bool
    },

    getDefaultProps: function () {
        return {
            disabled: false,
            tagName: "div"
        };
    },

    render: function () {
        var connectDragSource = this.props.connectDragSource;
        var connectDropTarget = this.props.connectDropTarget;

        var opacity = this.props.isDragging ? 0.2 : 1;

        var rowProps = {
            className: classnames("drag-drop-item", this.props.className),
            "data-id": "drag-drop-item",
            style: _.clone(_.extend(this.props.style, { opacity: opacity }))
        };

        var row = (
        React.createElement(this.props.tagName, rowProps, this.props.children));
        //IE must have a drop target even if dragging is disabled.  Enabling after the first render has no effect.
        return this.props.removeDraggableAttribute ? connectDropTarget(row) : connectDragSource(connectDropTarget(row));
    }
});

DragDrop = dropTarget(TYPE, dropSpec, dropCollect)(
    dragSource(TYPE, dragSpec, dragCollect)(DragDrop)
);

//expose these functions for unit tests
DragDrop.dropSpec = dropSpec;
DragDrop.dragSpec = dragSpec;
DragDrop.isInTopHalf = isInTopHalf;
DragDrop.isInLeftHalf = isInLeftHalf;

module.exports = DragDrop;
