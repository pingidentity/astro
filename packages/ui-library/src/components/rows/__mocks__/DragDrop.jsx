var React = require("react");

module.exports = React.createClass({
    render: function () {
        return (<div {...this.props}>{this.props.children}</div>);
    }
});
