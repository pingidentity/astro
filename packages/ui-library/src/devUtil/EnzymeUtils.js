import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
