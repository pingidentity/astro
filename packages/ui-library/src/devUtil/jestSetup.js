
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
