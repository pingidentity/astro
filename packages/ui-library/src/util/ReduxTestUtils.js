var React = require("react");

/**
 * @module util/ReduxTest
 * @desc The module contains utility functions to help with testing Redux componentn
 */

/**
 * @function createTestDispatcher
 * @desc When writing unit tests for a reducer, there needs to be a convenient way to induce the reducer and execute
 * the middleware.  This function will create a mock dispatcher for the closed over state
 * @param {function} reducer - The reducer to bind to the dispatch
 * @param {object} obj - The initial state wrapped in an object so we get a pointer to the
 * @returns {function} - Returns a dispatch function
 * @example:
 * var wrapper = { state: {} };
 * var dispatch = createTestDispatcher(reducer, wrapper);
 *
 * dispatch(Action.someAction());
 */
exports.createTestDispatcher = function (reducer, obj) {
    var dispatch = function (action) {
        if (typeof(action) === "function") {
            action(dispatch);
        } else {
            obj.state = reducer(obj.state, action);
        }
    };

    return dispatch;
};

/**
 * @class Wrapper
 * @desc Since react has deprecated using set props, we need to wrap components in another component to enable
 * sending properties to them by exposing a function.
 */
exports.Wrapper = React.createClass({
    sendProps: function (state) {
        this.setState(state);
    },

    render: function () {
        /* eslint-disable */
        return <this.props.type ref="target" {...this.props.opts} {...this.state} />;
        /* eslint-enable */
    }
});
