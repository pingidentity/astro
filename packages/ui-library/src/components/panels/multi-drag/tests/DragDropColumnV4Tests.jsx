window.__DEV__ = true;

jest.dontMock("../DragDropColumn");
jest.dontMock("../../../../util/EventUtils");
jest.dontMock("../../../../util/Utils");
jest.mock("popper.js");
jest.mock("react-portal");

import React from "react";
import ReactDOM from "react-dom";
import { mountSnapshotDataIds } from "../../../../devUtil/EnzymeUtils";
import TestUtils from "../../../../testutil/TestUtils";
import ReactTestUtils from "react-dom/test-utils";
import ReduxTestUtils from "../../../../util/ReduxTestUtils";
import DragDropColumn from "../DragDropColumn";
import DragDrop from "../../../rows/DragDrop";
import _ from "underscore";
import TestBackend from "react-dnd-test-backend";
import { DragDropContext } from "react-dnd";

describe("DragDropColumn v4", function () {

    //this is set for each component in wrapInTestContext ref. Keeps from chaning all of the string refs
    let thisComponent;

    function wrapInTestContext (Component) {
        return DragDropContext(TestBackend)( class extends React.Component {
            render() {
                return <Component ref={(c) => thisComponent = c} {...this.props} />;
            }
        });
    }

    function getOpts (opts) {
        return _.defaults(opts || {}, {
            onDrag: jest.fn(),
            onDrop: jest.fn(),
            onCancel: jest.fn(),
            onSearch: jest.fn(),
            onScrolledToTop: jest.fn(),
            onScrolledToBottom: jest.fn(),

            contentType: <div />,
            name: "Available Rows",
            index: 0,
            rows: [{ id: 1, n: 1 }, { id: 2, n: 2 }],

        });
    }

    function getComponent (opts) {
        opts = getOpts(opts);

        const WrappedComponent = wrapInTestContext(DragDropColumn);
        return ReactTestUtils.renderIntoDocument(<WrappedComponent {...opts} >Content</WrappedComponent>);
    }

    function getWrappedComponent (opts) {
        opts = getOpts(opts);

        const WrappedComponent = wrapInTestContext(DragDropColumn);
        return ReactTestUtils.renderIntoDocument(<ReduxTestUtils.Wrapper type={WrappedComponent} opts={opts} />);
    }

    beforeEach(function () {
        thisComponent = null;
    });

    it("data-id's don't change", () => {
        const WithContext = DragDropContext(TestBackend)(() => (
            <DragDropColumn
                onDrag={jest.fn()}
                onDrop={jest.fn()}
                onCancel={jest.fn()}
                onSearch={jest.fn()}
                onScrolledToTop={jest.fn()}
                onScrolledToBottom={jest.fn()}

                contentType={<div />}
                helpText="PLEASE HELP I AM LOST"
                name={"Available Rows"}
                index={0}
                rows={[{ id: 1, n: 1 }, { id: 2, n: 2 }]}

            />
        ));
        mountSnapshotDataIds(
            <WithContext />
        );
    });

    it("renders with default data-id", function () {
        getWrappedComponent();
        const component = thisComponent;

        const dragDropColumn = TestUtils.findRenderedDOMNodeWithDataId(component, "drag-drop-column");

        expect(dragDropColumn).toBeDefined();
    });

    it("renders with given data-id", function () {
        getWrappedComponent({ "data-id": "myDragDropColumn" });
        const component = thisComponent;

        const dragDropColumn = TestUtils.findRenderedDOMNodeWithDataId(component, "myDragDropColumn");

        expect(dragDropColumn).toBeDefined();
    });

    it("renders with given className", function () {
        getWrappedComponent({ className: "myDragDropColumnClass" });
        const component = thisComponent;

        const dragDropColumn = TestUtils.findRenderedDOMNodeWithClass(component, "myDragDropColumnClass");

        expect(dragDropColumn).toBeDefined();
    });

    it("Renders drop target when no rows", function () {
        getWrappedComponent({ rows: [] });
        const component = thisComponent;

        expect(TestUtils.findRenderedDOMNodeWithDataId(component, "empty-placeholder")).toBeTruthy();
    });



    it("renders ghost row", function () {
        getWrappedComponent({ ghostRowAt: 1 });
        const component = thisComponent;

        //find all the rendered drag drop rows
        const rows = ReactDOM.findDOMNode(component).getElementsByClassName("drag-drop-item");

        expect(rows.length).toBe(3);

        //TODO: reactid is deprecated
        //expect(rows[1].getAttribute("data-reactid")).toContain("preview");
    });


    it("triggers correct category list events", function() {
        const toggleCallback = jest.fn();
        const clickCallback = jest.fn();

        const component = getWrappedComponent({
            categoryList: ["One", "Two"],
            showCategoryList: true,
            onCategoryToggle: toggleCallback,
            onCategoryClick: clickCallback,
        });

        const categoryToggle = TestUtils.findRenderedDOMNodeWithDataId(
            component,
            "link-dropdown-list-label"
        );
        const categoryOption = TestUtils.scryRenderedDOMNodesWithClass(component, "select-option")[0];

        expect(toggleCallback).not.toBeCalled();
        ReactTestUtils.Simulate.click(categoryToggle);
        expect(toggleCallback).toBeCalled();

        expect(clickCallback).not.toBeCalled();
        ReactTestUtils.Simulate.click(categoryOption);
        expect(clickCallback).toBeCalled();
    });

    it("triggers onDrag callback", function() {
        const callback = jest.fn();
        const mockOffset = {
            clientOffset: { x: 0, y: 0 },
            getSourceClientOffset: function () { return { x: 0, y: 0 }; }
        };

        const root = getComponent({ onDrag: callback });

        const backend = root.getManager().getBackend();
        const dropzone = ReactTestUtils.scryRenderedComponentsWithType(root, DragDrop)[0];
        const draggable = dropzone.decoratedComponentInstance;

        expect(callback).not.toBeCalled();
        backend.simulateBeginDrag([draggable.getHandlerId()], mockOffset);
        backend.simulateHover([dropzone.getHandlerId()], mockOffset);
        backend.simulateDrop();
        expect(callback).toBeCalled();
    });

    it("doesn't trigger onDrag callback because sort is disabled", function() {
        const callback = jest.fn();
        const mockOffset = {
            clientOffset: { x: 0, y: 0 },
            getSourceClientOffset: function () { return { x: 0, y: 0 }; }
        };

        const root = getComponent({ onDrag: callback, disableSort: true });

        const backend = root.getManager().getBackend();
        const dropzone = ReactTestUtils.scryRenderedComponentsWithType(root, DragDrop)[0];
        const draggable = dropzone.decoratedComponentInstance;

        backend.simulateBeginDrag([draggable.getHandlerId()], mockOffset);
        backend.simulateHover([dropzone.getHandlerId()], mockOffset);
        backend.simulateDrop();
        expect(callback).not.toBeCalled();
    });

    it("renders the column title helphint", () => {
        const dataId = "testdataid";
        const name = "Test column name";
        const helpText = "Test help text!";

        getWrappedComponent({
            "data-id": dataId,
            helpText: helpText,
            name: name,
        });

        const component = thisComponent;

        const helpHintElement = TestUtils.findRenderedDOMNodeWithDataId(component, `${dataId}-helphint`);
        expect(helpHintElement).toBeTruthy();

        const helpHintTarget = TestUtils.findRenderedDOMNodeWithDataId(component, `${dataId}-helphint-target`);
        expect(helpHintTarget.className).toContain("row-selector__column-helptext inline");

        const helpHintText = TestUtils.findRenderedDOMNodeWithDataId(helpHintElement, "tooltip");
        expect(helpHintText.textContent).toEqual(helpText);
    });

    it("accepts render prop as contentType with correct onAdd", () => {
        const onAdd = jest.fn();
        getWrappedComponent({
            contentType: props => props.onAdd(),
            onAdd
        });

        expect(onAdd).toHaveBeenCalled();
    });

    it("Executes scroll callbacks", function () {
        getWrappedComponent({ rows: [] });
        const component = thisComponent;
        const itemNode = ReactDOM.findDOMNode(component.refs.items);

        //have to extend this because jsdom doesn't allow to set scrollTop etc
        const items = _.extend({}, itemNode);

        //mocking find dom node in the function to return our new items instead
        ReactDOM.findDOMNode = jest.fn().mockReturnValue(items);
        items.getBoundingClientRect = jest.fn().mockReturnValue({
            height: 100
        });

        component._handleScroll();
        expect(component.props.onScrolledToTop).toBeCalled();
        expect(component.props.onScrolledToBottom).not.toBeCalled();
        //clear the mock
        component.props.onScrolledToTop.mockClear();

        //set up the node to look like it's scrolled to the bottom
        // we're overshooting because it seems like in practice, that can happen
        items.scrollTop = 105;
        items.scrollHeight = 200;

        component._handleScroll();
        expect(component.props.onScrolledToTop).not.toBeCalled();
        expect(component.props.onScrolledToBottom).toBeCalled();
        //clear the mock
        component.props.onScrolledToBottom.mockClear();

        //set up the node to look like it's scrolling back up
        items.scrollTop = 100;
        items.scrollHeight = 200;

        component._handleScroll();
        expect(component.props.onScrolledToBottom).not.toBeCalled();

        jest.restoreAllMocks();
    });
    // *** Make sure the above test is always last because of the mock findDOMNode ***

});
