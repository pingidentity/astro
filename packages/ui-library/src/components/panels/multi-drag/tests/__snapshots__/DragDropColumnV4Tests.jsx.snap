// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragDropColumn v4 data-id's don't change 1`] = `
Array [
  "drag-drop-column",
  "drag-drop-column-helphint",
  "drag-drop-column-helphint-target",
  "drag-drop-column-helphint-icon",
  "tooltip",
  "drag-drop-row_1",
  "row_1",
  "drag-drop-row_2",
  "row_2",
]
`;
